dftimewolf/cli/curses_display_manager.py

```diff
@@ -0,0 +1,290 @@
+    """Enqueue a preflight module object for display."""
+    m = Module(name, dependencies, runtime_name)
+    self._preflights[m.runtime_name] = m
+
+  def EnqueueModule(self,
+                       name: str,
+                       dependencies: List[str],
+                       runtime_name: Optional[str]) -> None:
+    """Enqueue a module object for display."""
+    m = Module(name, dependencies, runtime_name)
+    self._modules[m.runtime_name] = m
+
+  def UpdateModuleStatus(self, module: str, status: Status) -> None:
+    """Update the status of a module for display."""
+    if module in self._preflights:
+      self._preflights[module].SetStatus(status)
+    if module in self._modules:
+      self._modules[module].SetStatus(status)
+
+    self.Draw()
+
+  def SetThreadedModuleContainerCount(self, module: str, count: int) -> None:
+    """Set the container count that a threaded module will operate on."""
+    if module in self._preflights:
+      self._preflights[module].SetContainerCount(count)
+    if module in self._modules:
+      self._modules[module].SetContainerCount(count)
+
+  def UpdateModuleThreadState(self,
+                              module: str,
+                              status: Status,
+                              thread: str,
+                              container: str) -> None:
+    """Update the state of a thread within a threaded module for display."""
+    if module in self._preflights:
+      self._preflights[module].SetThreadState(thread, status, container)
+    if module in self._modules:
+      self._modules[module].SetThreadState(thread, status, container)
+
+    self.Draw()
+
+  def Draw(self) -> None:
+    """Draws the window."""
+    if not self._stdscr:
+      return
+
+    with self._lock:
+      self._stdscr.clear()
+      y, _ = self._stdscr.getmaxyx()
+
+      curr_line = 1
+      self._stdscr.addstr(curr_line, 0, f' {self._recipe_name}')
+      curr_line += 1
+
+      # Preflights
+      if self._preflights:
+        self._stdscr.addstr(curr_line, 0, '   Preflights:')
+        curr_line += 1
+        for _, module in self._preflights.items():
+          for line in module.Stringify():
+            self._stdscr.addstr(curr_line, 0, line)
+            curr_line += 1
+
+      # Modules
+      self._stdscr.addstr(curr_line, 0, '   Modules:')
+      curr_line += 1
+      for status in Status:  # Print the modules in Status order
+        for _, module in self._modules.items():
+          if module.status != status:
+            continue
+          for line in module.Stringify():
+            self._stdscr.addstr(curr_line, 0, line)
+            curr_line += 1
+
+      # Messages
+      curr_line += 1
+      self._stdscr.addstr(curr_line, 0, ' Messages:')
+      curr_line += 1
+
+      message_space = y - 4 - curr_line
+      start = len(self._messages) - message_space
+      start = 0 if start < 0 else start
+
+      for m in self._messages[start::]:
```

Comment:
Why the `::` here? If we need to create a slice copy, it would be nice to add a comment detailing why.

Reply:
A relic of an earlier design, where I had the messages printed in reverse order. Fixed.